Implement functionality to conveniently watch only a single path.

magJ opened the pull request:
Modified the PathWatcher start method to return itself to facilitate call chaining.
Changed the contract and implementation of the onStart lifecycle so that onStart is called once the watcher has actually started working.

@danielmitterdorfer I sort of just tacked on most of this functionality as I didn't want to interfere too much with the existing API.
But there is probably a better way of structuring it. In particular the `SinglePathChangeListenerAdapter` just undoes the work of the existing `WatchServicePathWatcher#handleEvent` method.  

There is also a potential problem with the change I did to the onStart call, since it will return the instance of the inner `WatchServicePathWatcher` rather than the `RunnablePathWatcherAdapter` that it used to.

magJ committed:
Implement functionality to conveniently watch only a single path.
Modified the PathWatcher start method to return itself to facilitate call chaining.
Changed the contract and implementation of the onStart lifecycle so that onStart is called once the watcher has actually started working.

danielmitterdorfer commented:
Hi @magJ, first of all thanks for your PR! Unfortunately, I have a very busy week but I try to review your PR as soon as I can. Just wanted to let you know that it will probably take a few days until I can look more closely into it.

magJ commented:
@danielmitterdorfer No rush.

danielmitterdorfer commented:
I started to look at your PR now a bit more closely. First of all I think this adds useful functionality but I think it would be great to have it a bit more tightly integrated with Perlock.

I also think that one case should be clearly documented: If the watched file is renamed, the watcher will not produce any more events.

I will try to set aside time tomorrow in the morning to add more concrete comments.

danielmitterdorfer commented on examples/perlock-single-simple-demo/src/main/java/name/mitterdorfer/perlock/examples/PathWatcherDemo.java:
```diff
@@ -0,0 +1,44 @@
+ * parameters:
+ *
+ * <ol>
+ *     <li><code>--recursive</code> to watch a path recursively or <code>--non-recursive</code> to watch
```
Copy & Paste Error?

danielmitterdorfer commented on examples/perlock-single-simple-demo/src/main/java/name/mitterdorfer/perlock/examples/PathWatcherDemo.java:
```diff
@@ -0,0 +1,44 @@
+ *      <li><code>path</code> which is the path to watch</li>
+ * </ol>
+ *
+ * If you invoke the sample application as follows, it watches the home folder (on Unix): <code>java -jar perlock-examples.jar --non-recursive ~</code>
```
Copy & Paste Error?

danielmitterdorfer commented on examples/perlock-single-simple-demo/src/main/java/name/mitterdorfer/perlock/examples/PathWatcherDemo.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * If you invoke the sample application as follows, it watches the home folder (on Unix): <code>java -jar perlock-examples.jar --non-recursive ~</code>
+ */
+class SinglePathWatcherDemo {
```
I think the class name should match the file name. Is it intentional that the class is not declared `public`?

magJ replied:
Class name is a basic error and the package-private scope was the reason I didn't notice it.

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/PathWatcher.java:
```diff
@@ -17,12 +17,12 @@
      * Determines whether this <code>PathWatcher</code> instance is currently running.
      *
-     * @return true iff this <code>PathWatcher</code> instance is currently running.
+     * @return true if this <code>PathWatcher</code> instance is currently running.
```
the "iff" was intentional. It means "if and only if" and is quite common (see [Wikipedia](https://en.wikipedia.org/wiki/If_and_only_if))

magJ replied:
Oh, I didn't know, thanks.

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/SinglePathChangeListener.java:
```diff
@@ -0,0 +1,18 @@
+ * thread. This implies that callback methods should return reasonably fast and offload heavy lifting to a dedicated
+ * thread. Otherwise, events may be lost.</p>
+ */
+public interface SinglePathChangeListener {
```
I think this interface is great and better than the existing `PathChangeListener` (especially since it can be used as a functional interface in Java 8+). Would you be willing you replace `PathChangeListener` with your interface in general in Perlock? I think it is fine as Perlock is at the moment in version 0.3 and we can introduce breaking API changes.

This would also eliminate the need for `AbstractPathChangeListener`.

magJ replied:
I will take a look at making that change.

magJ replied:
So I will modify the `PathChangeListener` to have the method

``` java
void onPathChanged(WatchEvent.Kind<Path> eventKind, Path path);
```

This is a hybrid between the two interfaces, as the `SingleChangePathListener` does not provide the `Path` since it is implicit.

danielmitterdorfer replied:
Sounds good!

magJ replied:
Thinking about it some more has sort of opened a can of worms. Perhaps the API should just be 

``` java
void onPathChanged(WatchEvent<Path> event);
```

We could potentially just return the original `WatchEvent`, however this would mean that the `context()` would return a relative `Path`.
One option could be implementing `WatchEvent` and setting the `context()` to a resolved path, but this would be against the `WatchEvent` documentation.
One more option would be to expose our own version of a watch event, like the one you have at `AssertingPathChangeListener.Event` that is a bit cleaner to work with.

One other thing that might be worth considering from an API point of view is `OVERFLOW` events.
Currently perlock silently ignores them, but I think there should be some way for the user to be made aware of these.

Perhaps the API should just provide a `WatchEvent<?>` and let the user deal with casting and resolving the path.

danielmitterdorfer replied:
Oh man, sorry for that. I have maybe time to look more closely into this but the most viable option seems to be a custom event class (this would also decouple the interface a bit from `WatchService` although I wouldn't worry too much about this here).

> One other thing that might be worth considering from an API point of view is OVERFLOW events.
> Currently perlock silently ignores them, but I think there should be some way for the user to be made aware of these.

Good suggestion. But I'd say this is a separate issue.

danielmitterdorfer replied:
I have now opened #2 for tracking the OVERFLOW events topic.

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/SinglePathWatcherFactory.java:
```diff
@@ -0,0 +1,107 @@
+ * Factory to create {@link PathWatcher}s that only deal with a single file and
+ * ignore change to other files within the directory
+ */
+public class SinglePathWatcherFactory {
```
I think a separate factory is a fairly large change. I understand that you did not want to change the existing API but I think overall it would be better to merge this class with `PathWatcherFactory`.

What about a new public method `createSinglePathWatcher()` in `PathWatcherFactory`? It should also have a precondition check that a `Path` representing a file is provided.

magJ replied:
I can move these into the `PathWatcherFactory`.
In regards to checking that a `Path` represents a file, do you mean checking if `Files#isRegularFile`?
I am not sure that check is necessary or even desirable, it is conceiveable that the user may want to monitor the creation and deletion of a particular directory. So long as the path is non-null I think it would suffice.

danielmitterdorfer replied:
If the user calls the method `createSinglePathWatcher()` (which is intended to watch only a single file) I think the precondition `Files#isRegularFile() == true` must hold for the provided path. Otherwise they have to call another create method that is able to watch a directory, no?

magJ replied:
The user may want to know when `mydirectory` IE `/rootPath/mydirectory/` is created or deleted.
They may not be at all interested in the content of  `mydirectory`.

magJ replied:
Also, `isRegularFile` will return false if the file does not exist.
The user will not be able to watch for file creation unless they create and delete the file before and after starting the path watcher.

danielmitterdorfer replied:
I got you now. This makes sense. And yes, `isRegularFile` will return `false` for non-existing files. However, I thought this was just intended for watching for changes in existing files but obviously this won't work when watching for file creation. So just disregard my comments related to this check. :)

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/impl/WatchServicePathWatcher.java:
```diff
@@ -75,11 +83,13 @@ private void performRegistration() throws IOException {
         prepareWatcher();
         performRegistration();
         running = true;
+        lifecycleListener.onStart(this);
```
I think this line is problematic for multiple reasons:
1. We have no control whatsoever over the lifecycle listener, meaning that it could throw an exception due to a bug in the implementation of the lifecycle listener and we don't detect it (by default the thread will just terminate and that's it but you'll never see the exception). That's why the original code used the `Silently.run()` idiom.
2. By providing `this` to the lifecycle listener we're leaking an implementation detail to the listener. I think this can be mitigated by creating a wrapper lifecycle listener that will then replace the sender with the outer class (i.e. `RunnablePathWatcherAdapter`). I hope you know how I mean it. :)

magJ replied:
Regarding 1, that is a good point that I missed, I will address it.

Regarding 2 I'm not sure exactly what the right approach would be, but I will take another look.

The important thing is that the user has a guaranteed way of knowing when the watch service has begun watching for events.
One thing I considered after submitting this PR was having the `RunnablePathWatcherAdapter#start()` method block until it actually starts.
I think it would be ideal to implement both, the accurate life-cycle and the start method blocking.

danielmitterdorfer replied:
> One thing I considered after submitting this PR was having the RunnablePathWatcherAdapter#start() method block until it actually starts.

I just thought about it. I think the easiest way to achieve this is by splitting the implementation in `WatchServicePathWatcher#start()`. `#start()` would just call `prepareWatcher()`and `performRegistration()` and set the flag `running` to `true`.

The adapter could then invoke the `start()` method synchronously:

```
public void start() {
            if (future != null) {
                throw new IllegalStateException("Cannot start a PathWatcher that is already running.");
            }
           // synchronous start
           delegate.start();
            LOG.trace("Submitting '{}' to executor service.", delegate);
            //submit itself when client wants to start watching. The pool will invoke the runnable when its ready
            future = executorService.submit(this);
        }
```

in the constructor of `RunnablePathWatcherAdapter` you have to store the concrete implementation class `WatchServicePathWatcher` as a field and in its `run()` method you just call `delegate.watch()` instead of `#start()`.

danielmitterdorfer commented on perlock-core/src/test/java/name/mitterdorfer/perlock/SinglePathWatcherFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+        assertEquals(ENTRY_MODIFY, queue.poll(testTimeout, TimeUnit.SECONDS));
+
+        Files.delete(filepath);
+        assertEquals(ENTRY_DELETE, queue.poll(testTimeout, TimeUnit.SECONDS));
```
I think there are a few interesting cases that you do not test:
- Adding / changing another file in the same directory does not fire a event (as that's what your change is about :))
- After the file has been renamed you get no further events.

magJ replied:
I will add these test cases.

danielmitterdorfer replied:
Thanks!

danielmitterdorfer commented:
I added a couple of suggestions. Just let me know what you think. In general I think that this is a useful feature!

magJ committed:
Incorporated review feedback.
Upgraded jimfs dependency so that polling time is configurable, significantly improving unit test speed.
Refactored the PathChangeListener interface to make it suited for use as a functional interface.
Refactored the RunnablePathWatcherAdapter  adapter so that starting the pathwatcher will block until the watcher is listening.
Added some more test cases for the single path watcher functionality.

magJ commented:
@danielmitterdorfer I have incorporated your feedback.
Please take another look when you get a chance.

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/EventKind.java:
```diff
@@ -0,0 +1,45 @@
+        WATCH_EVENT_KIND_TO_EVENT_KIND = Collections.unmodifiableMap(eventKindMap);
+    }
+
+    private WatchEvent.Kind<Path> watchEventKind;
```
Can you declare this as `final`?

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/PathChangeListener.java:
```diff
@@ -3,32 +3,19 @@
     /**
      * This method has to be called when a new path is created.
      *
      * @param path The path that has been created. Must not be null.
```
The Javadoc seems to be outdated now.

danielmitterdorfer commented on perlock-core/src/main/java/name/mitterdorfer/perlock/impl/WatchServicePathWatcher.java:
```diff
@@ -87,8 +83,7 @@ public PathWatcher start() throws IOException {
         running = true;
-        lifecycleListener.onStart(this);
-        watch();
+        //watch();
```
Leftover?

danielmitterdorfer commented on perlock-core/src/test/java/name/mitterdorfer/perlock/PathWatcherFactoryTest.java:
```diff
@@ -29,7 +31,8 @@
     @Before
     public void setUp() throws Exception {
-        FileSystem fs = Jimfs.newFileSystem(Configuration.unix());
+        WatchServiceConfiguration wsc = WatchServiceConfiguration.polling(100, TimeUnit.MILLISECONDS);
```
It's great that JimFs allows now shorter polling cycles. :)

danielmitterdorfer commented:
@magJ I left a few very minor comments but apart from that it looks very good. Thanks for tackling this.

magJ committed:
Incorporated review feedback.
Fixed javadoc comment.
Removed commented out code.
Made enum field variable final.

magJ commented:
Thanks @danielmitterdorfer feedback incorporated.

danielmitterdorfer commented:
Thanks a lot for your PR @magJ! I'll merge it soon and will release a new version of Perlock when I have a little bit of time to spare (probably over the weekend).

danielmitterdorfer commented:
@magJ I just released Perlock 0.3.1. The artifacts are not yet synced to Maven central but should be available soon.

magJ commented:
Thanks @danielmitterdorfer
